crate/memories/src/batching_layer/buffer.rs

```diff
@@ -0,0 +1,95 @@
+use std::{mem, sync::Mutex};
+
+use crate::{
+    BatchingMemoryADT,
+    batching_layer::operation::{Operation, PendingOperations},
+};
+
+struct Buffer<M: BatchingMemoryADT> {
+    capacity: usize, // the size at which the buffer should be flushed
+    data: PendingOperations<M>,
+}
+
+impl<M: BatchingMemoryADT> Buffer<M> {
+    /// Flushes the buffer if it contains data and returns the flushed
+    /// operations. Returns None if the buffer is empty.
+    fn flush_if_not_empty(&mut self) -> Option<PendingOperations<M>> {
+        if !self.data.is_empty() {
+            Some(mem::take(&mut self.data))
+        } else {
+            None
+        }
+    }
+}
+
+pub(crate) struct ThreadSafeBuffer<M: BatchingMemoryADT>(Mutex<Buffer<M>>);
+
+impl<M> ThreadSafeBuffer<M>
+where
+    M: BatchingMemoryADT,
+{
+    pub(crate) fn new(capacity: usize) -> Self {
+        Self(Mutex::new(Buffer::<M> {
+            capacity,
+            data: Vec::with_capacity(capacity),
+        }))
+    }
+
+    pub(crate) fn shrink_capacity(&self) -> Result<Option<PendingOperations<M>>, BufferError> {
+        let mut buffer = self.0.lock()?;
+        if buffer.capacity == 0 {
+            return Err(BufferError::Underflow);
+        }
+        buffer.capacity -= 1;
+        Ok(buffer.flush_if_not_empty())
+    }
+
+    pub(crate) fn push(
+        &self,
+        item: Operation<M>,
+    ) -> Result<Option<PendingOperations<M>>, BufferError> {
+        let mut buffer = self.0.lock()?;
+        // Check if the new item is compatible with the last item, since the buffer is
+        // thread-safe, this ensures by transitivity that all items in the
+        // buffer are of the same type.
+        if let Some(last_item) = buffer.data.last() {
+            if mem::discriminant(last_item) != mem::discriminant(&item) {
+                return Err(BufferError::TypeMismatch);
+            }
+        }
+        buffer.data.push(item);
+        Ok(buffer.flush_if_not_empty())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BufferError {
+    TypeMismatch, // when the type of the new item does not match the type of the last item
+    Overflow,
+    Underflow,
+    Mutex(String),
+}
+
+impl<M: BatchingMemoryADT> From<std::sync::PoisonError<std::sync::MutexGuard<'_, Buffer<M>>>>
+    for BufferError
+{
+    fn from(e: std::sync::PoisonError<std::sync::MutexGuard<'_, Buffer<M>>>) -> Self {
+        Self::Mutex(format!("Mutex poisoned: {}", e))
+    }
+}
```

Comment:
Simply unwrap upon lock poisoning:
```suggestion
}
```

Reply:
okay DONE... but why ?

we will probably face some clippy consequences (that I will silence, for the matter)